### ISSUE TITLE
Custom error in history api

### ISSUE THREAD

**Issue by Kl4rry** [kkawakam/rustyline#762]
Currently the history trait requires the error return to be a readline error. That means that one cannot return other error types that are not included in:
```
pub enum ReadlineError {
    Io(Error),
    Eof,
    Interrupted,
    Errno(Error),
    WindowResized,
}
```
This can be fixed by either adding a `Box<dyn std::error::Error>` variant that can be  used to return any `std::error::Error` or by adding a optional generic error type like this:
```
pub enum ReadlineError<E = std::convert::Infallible> {
    Io(Error),
    Eof,
    Interrupted,
    Errno(Error),
    WindowResized,
    UserError(E),
}
```

**Comment by Kl4rry** [comment]
Another possible solution is to make `History` generic over the error type and having a trait bound that it must implement into for `ReadlineError`.

**Comment by gwenn** [comment]
Just for reference:
- https://docs.rs/termwiz/latest/termwiz/lineedit/trait.History.html (no error)
- https://docs.rs/reedline/latest/reedline/trait.History.html (only https://docs.rs/reedline/latest/reedline/enum.ReedlineErrorVariants.html)

As far as I know, other rust libraries doesn't support custom `History` impl.

**Comment by Kl4rry** [comment]
Termwiz kinda supports every error type as you can just to `.ok()` on every error type.
There is a work around that can be used in both reedline and rustyline and that is to put the error in a IO error like this
`std::io::Error::new(std::io::ErrorKind::InvalidData, Box::new(custom_error))`

**Comment by yamamel** [comment]
I agree with this. There should be a case to allow user return other errors they encountered.
